File: api_gateway/src/core/service/logistics/inventory_item_service.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package service

import (
	"context"

	"github.com/golibs-starter/golib/log"
	"github.com/serp/api-gateway/src/core/domain/dto/response"
	port "github.com/serp/api-gateway/src/core/port/client/logistics"
)

type IInventoryItemService interface {
	CreateInventoryItem(ctx context.Context, req map[string]interface{}) (*response.BaseResponse, error)
	UpdateInventoryItem(ctx context.Context, inventoryItemId string, req map[string]interface{}) (*response.BaseResponse, error)
	DeleteInventoryItem(ctx context.Context, inventoryItemId string) (*response.BaseResponse, error)
	GetInventoryItem(ctx context.Context, inventoryItemId string) (*response.BaseResponse, error)
	GetInventoryItems(ctx context.Context, page, size int, sortBy, sortDirection, query, productId, facilityId, expirationDateFrom, expirationDateTo, manufacturingDateFrom, manufacturingDateTo, statusId string) (*response.BaseResponse, error)
}

type InventoryItemService struct {
	inventoryItemClient port.IInventoryItemClientPort
}

func (i *InventoryItemService) CreateInventoryItem(ctx context.Context, req map[string]interface{}) (*response.BaseResponse, error) {
	res, err := i.inventoryItemClient.CreateInventoryItem(ctx, req)
	if err != nil {
		log.Error(ctx, "InventoryItemService: CreateInventoryItem error: ", err.Error())
		return nil, err
	}
	return res, nil
}

func (i *InventoryItemService) UpdateInventoryItem(ctx context.Context, inventoryItemId string, req map[string]interface{}) (*response.BaseResponse, error) {
	res, err := i.inventoryItemClient.UpdateInventoryItem(ctx, inventoryItemId, req)
	if err != nil {
		log.Error(ctx, "InventoryItemService: UpdateInventoryItem error: ", err.Error())
		return nil, err
	}
	return res, nil
}

func (i *InventoryItemService) DeleteInventoryItem(ctx context.Context, inventoryItemId string) (*response.BaseResponse, error) {
	res, err := i.inventoryItemClient.DeleteInventoryItem(ctx, inventoryItemId)
	if err != nil {
		log.Error(ctx, "InventoryItemService: DeleteInventoryItem error: ", err.Error())
		return nil, err
	}
	return res, nil
}

func (i *InventoryItemService) GetInventoryItem(ctx context.Context, inventoryItemId string) (*response.BaseResponse, error) {
	res, err := i.inventoryItemClient.GetInventoryItem(ctx, inventoryItemId)
	if err != nil {
		log.Error(ctx, "InventoryItemService: GetInventoryItem error: ", err.Error())
		return nil, err
	}
	return res, nil
}

func (i *InventoryItemService) GetInventoryItems(ctx context.Context, page, size int, sortBy, sortDirection, query, productId, facilityId, expirationDateFrom, expirationDateTo, manufacturingDateFrom, manufacturingDateTo, statusId string) (*response.BaseResponse, error) {
	res, err := i.inventoryItemClient.GetInventoryItems(ctx, page, size, sortBy, sortDirection, query, productId, facilityId, expirationDateFrom, expirationDateTo, manufacturingDateFrom, manufacturingDateTo, statusId)
	if err != nil {
		log.Error(ctx, "InventoryItemService: GetInventoryItems error: ", err.Error())
		return nil, err
	}
	return res, nil
}

func NewInventoryItemService(inventoryItemClient port.IInventoryItemClientPort) IInventoryItemService {
	return &InventoryItemService{
		inventoryItemClient: inventoryItemClient,
	}
}

File: sales/src/main/java/serp/project/sales/service/InventoryItemService.java
package serp.project.sales.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import serp.project.sales.dto.request.InventoryItemCreationForm;
import serp.project.sales.dto.request.InventoryItemUpdateForm;
import serp.project.sales.entity.InventoryItemEntity;
import serp.project.sales.exception.AppErrorCode;
import serp.project.sales.exception.AppException;
import serp.project.sales.repository.InventoryItemRepository;
import serp.project.sales.repository.specification.InventoryItemSpecification;
import serp.project.sales.util.PaginationUtils;

import java.time.LocalDate;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional(rollbackFor = Exception.class)
public class InventoryItemService {

    private final InventoryItemRepository inventoryItemRepository;

    public void createInventoryItem(InventoryItemCreationForm form, Long tenantId) {
        InventoryItemEntity inventoryItem = new InventoryItemEntity(form, tenantId);
        inventoryItemRepository.save(inventoryItem);
        log.info("[InventoryItemService] Created inventory item with ID {} for product ID: {}", inventoryItem.getId(),
                form.getProductId());
    }

    public void updateInventoryItem(String id, InventoryItemUpdateForm form, Long tenantId) {
        InventoryItemEntity inventoryItem = getInventoryItem(id, tenantId);
        if (inventoryItem == null) {
            throw new AppException(AppErrorCode.NOT_FOUND);
        }

        inventoryItem.update(form);
        inventoryItemRepository.save(inventoryItem);
        log.info("[InventoryItemService] Updated inventory item with ID {}", id);
    }

    public void deleteInventoryItem(String id, Long tenantId) {
        InventoryItemEntity inventoryItem = getInventoryItem(id, tenantId);
        if (inventoryItem == null) {
            throw new AppException(AppErrorCode.NOT_FOUND);
        }

        inventoryItem.cleanup();

        inventoryItemRepository.delete(inventoryItem);
        log.info("[InventoryItemService] Deleted inventory item with ID {}", id);
    }

    public InventoryItemEntity getInventoryItem(String id, Long tenantId) {
        InventoryItemEntity inventoryItem = inventoryItemRepository.findById(id).orElse(null);
        if (inventoryItem == null || !inventoryItem.getTenantId().equals(tenantId)) {
            log.info("[InventoryItemService] Inventory item with ID {} not found or tenant ID mismatch", id);
            return null;
        }
        return inventoryItem;
    }

    public Page<InventoryItemEntity> getInventoryItems(
            String query,
            String productId,
            String facilityId,
            LocalDate expirationDateFrom,
            LocalDate expirationDateTo,
            LocalDate manufacturingDateFrom,
            LocalDate manufacturingDateTo,
            String statusId,
            Long tenantId,
            int page,
            int size,
            String sortBy,
            String sortDirection) {
        Pageable pageable = PaginationUtils.createPageable(page, size, sortBy, sortDirection);
        return inventoryItemRepository.findAll(
                InventoryItemSpecification.satisfy(
                        query,
                        productId,
                        facilityId,
                        expirationDateFrom,
                        expirationDateTo,
                        manufacturingDateFrom,
                        manufacturingDateTo,
                        statusId,
                        tenantId),
                pageable);
    }

}
